core/src/main/scala/org/apache/spark/util/collection/SizeTrackingVector.scala

```diff
@@ -41,6 +41,6 @@ private[spark] class SizeTrackingVector[T: ClassTag]
    * Return a trimmed version of the underlying array.
    */
   def toArray: Array[T] = {
-    super.iterator.toArray
+    super.copyArrayWithLength(size)
```

Comment:
Hm, could this method also just be in the parent? it doesn't seem specific to the subclass even.